### PR TITLE
Fix pylint error torch.tensor is not callable

### LINKED ISSUE #24807: Pylint Error `torch.tensor is not callable`
## 🐛 Bug

Pylint returns the error `torch.tensor is not callable`

## To Reproduce

```python
import torch

if __name__ == "__main__":
    t = torch.tensor([1, 2, 3], dtype=torch.float64)
```

```
>>> pipenv run pylint test.py
test.py:4:8: E1102: torch.tensor is not callable (not-callable)
```

## Expected behavior

This shall not raise pylint error

## Environment

```
PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.2 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti

Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.1

Versions of relevant libraries:
[pip3] numpy==1.17.0
[pip3] torch==1.2.0
[conda] Could not collect

Pytorch installed in Pipenv Environment
pylint==2.3.1
```

## Additional context

None

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @bhosmer @smessmer @ljk53 @ailzhang @malfet @rgommers @xuzhao9 @gramster

### PR DESCRIPTION
**Pull request by pmeier**
Fixes #24807 and supersedes the stale #25093 (Cc @Microsheep). If you now run the reproduction

```python
import torch

if __name__ == "__main__":
    t = torch.tensor([1, 2, 3], dtype=torch.float64)
```

with `pylint==2.6.0`, you get the following output

```
test_pylint.py:1:0: C0114: Missing module docstring (missing-module-docstring)        
test_pylint.py:4:8: E1101: Module 'torch' has no 'tensor' member; maybe 'Tensor'? (no-
member)                                                                               
test_pylint.py:4:38: E1101: Module 'torch' has no 'float64' member (no-member) 
```

Now `pylint` doesn't recognize `torch.tensor` at all, but it is promoted in the stub. Given that it also doesn't recognize `torch.float64`, I think fixing this is out of scope of this PR.

---

## TL;DR

This BC-breaking only for users that rely on unintended behavior. Since `torch/__init__.py` loaded `torch/tensor.py` it was populated in `sys.modules`. `torch/__init__.py` then overwrote `torch.tensor` with the actual function. With this `import torch.tensor as tensor` does not fail, but returns the function rather than the module. Users that rely on this import need to change it to `from torch import tensor`.

### CODE REVIEW

**Review by ezyang — APPROVED** [review]
hurrah! Let's pray it doesn't break anything internally either

**Comment by ezyang** [comment]
Broke some internal imports, gonna have to fix 'em

**Comment by ezyang** [comment]
Oh boy, this is wild. This does break people, but the code that is broken is "wrong". With a `torch.tensor` module, `import torch.tensor as tensor` is valid and will give you the NON-module tensor function in scope as tensor. The existence of torch.tensor module deceives module loading into allowing it, but you don't actually end up loading the module.

I did a quick GitHub code search and it looks like this will break people, so maybe we need to do a deprecation cycle on this. https://github.com/search?l=Python&q=%22import+torch.tensor%22&type=Code

**Comment by pmeier** [comment]
This is indeed interesting. I've run 

```python
try:
    import torch.tensor as tensor
    print(type(tensor))
except Exception as error:
    print(error)

try:
    from torch import tensor
    print(type(tensor))
except Exception as error:
    print(error)
```

against 4 configurations:

1. PyTorch stable
    ```
    <class 'builtin_function_or_method'>
    <class 'builtin_function_or_method'>
    ```
2. PyTorch nightly
    ```
    <class 'builtin_function_or_method'>
    <class 'builtin_function_or_method'>
    ```
3. This PR as is
    ```
    No module named 'torch.tensor'
    <class 'builtin_function_or_method'>
    ```
4. This PR with an additional, empty `tensor.py`
    ```
    <class 'module'>
    <class 'module'>
    ```

As you can see from 3. the `import torch.tensor as tensor` would be broken after this PR. But from 4. we can see that this can't be caused solely by the `tensor.py` module. I'll investigate further.

**Comment by pmeier** [comment]
Something is special about the way `torch` is set up now. If a create a minimal project with a `tensor.py` module as well as a `tensor()` function in `__init__.py` I will only ever import the module, i. e. case 4. from before.

**Comment by pmeier** [comment]
So this works because in `__init__.py` we are importing `tensor.py` 

https://github.com/pytorch/pytorch/blob/86861095fa0644a98e916ef770861cd312b93935/torch/__init__.py#L447

and only afterwards populate `torch.tensor` with the function

https://github.com/pytorch/pytorch/blob/86861095fa0644a98e916ef770861cd312b93935/torch/__init__.py#L550-L554

Thus, after `__init__` is run, `sys.modules` contains the `torch.tensor` entry. If we then run `import torch.tensor as tensor` instead of importing this simply returns the attribute `tensor` from `torch` which is the function.

---

We maybe could give the user a hint that what he is doing is "wrong" by implementing an [import hook](https://docs.python.org/3.6/reference/import.html#import-hooks) that emits a warning whenever `torch.tensor` is imported. @ezyang Should I send a proposal?

**Comment by pmeier** [comment]
Implementing an import hook is actually quite easy:

```python
import sys
import warnings

class _Finder(importlib.abc.MetaPathFinder):
    def find_spec(self, fullname, path, target = None):
        if fullname == "torch.tensor":
            warnings.warn("You are importing wrongly!")


sys.meta_path.insert(0, _Finder())
```

Unfortunately it won't give us what we need:

1. If we **do not** import a dummy `torch.tensor` in `__init__`, a later `import torch.tensor` will trigger the warning but actually perform the import and overwrite the function.
2. If we **do** import a dummy, `import torch.tensor` will give you the function, but no warning will be emitted. This happens, because the `_Finder` will only be used if the module is not in `sys.modules`. 

---

To achieve what we want we could patch `sys.modules`. Since `dict.__getitem__` is read-only, we need to replace it with a lightweight wrapper class:

```python
import sys


class _Modules(dict):
    def __getitem__(self, item):
        if item == "torch.tensor":
            warnings.warn("You are importing wrongly!")
        return super().__getitem__(item)


sys.modules = _Modules(sys.modules)
```

This snippet must be placed before we import the `tensor` function.

**Comment by pmeier** [comment]
Instead of patching `sys.modules`, I've opted to insert a module mock for `torch.tensor`. If called, the mock triggers the warning and dispatches the call to the function `torch.tensor`.

---

In theory this should work just fine, but in practice it is not. Consider the snippet

```python
import torch
import sys
import importlib

from torch import tensor
print(tensor)

import torch.tensor as tensor
print(tensor)

print(getattr(sys.modules["torch"], "tensor"))

print(sys.modules["torch.tensor"])
print(importlib.import_module("torch.tensor"))
print(__import__("torch.tensor", globals(), locals(), [None], 0))
```

```
<built-in method tensor of type object at 0x7ff0e8d6e7a0>
<built-in method tensor of type object at 0x7ff0e8d6e7a0>
<built-in method tensor of type object at 0x7ff0e8d6e7a0>
<torch._TensorModuleWarner object at 0x7fef46bdec70>
<torch._TensorModuleWarner object at 0x7fef46bdec70>
<torch._TensorModuleWarner object at 0x7fef46bdec70>
```

For some reason the `import` statement does **not** return the content of `sys.modules`, but rather behaves like `getattr(sys.modules["torch"], "tensor")`. This is opposed to the behavior of `importlib` and `__import__`. 

I'm not sure if this intended but undocumented behavior or simply a bug. Anyway, I don't think there is currently a way to trigger a warning if a user is importing wrong without breaking BC. @ezyang How do you want to move forward?

**Comment by ezyang** [comment]
@pmeier Thanks for investigating this, and in particular seeing if we could do a proper deprecation cycle.  It doesn't look like there is an easy way, so I'm going to suggest we wait for the branch cut, and then just break people directly by landing the original version of this PR. I'll start submitting a diff internally to fix the internal broken sites.

I'm going on parental leave soon so we're going to have to find someone else to actually land this diff.

**Comment by pmeier** [comment]
> I'm going on parental leave soon so we're going to have to find someone else to actually land this diff.

In that case, could you ping / assign someone else?

**Comment by pmeier** [comment]
As a last comment, the behavior is intended and [documented](https://docs.python.org/3/reference/import.html#submodules):

> [...] if you have `sys.modules['spam']` and `sys.modules['spam.foo']` [...], the latter must appear as the foo attribute of the former.

In contrast, [`importlib.import_module` ](https://docs.python.org/3/library/importlib.html#approximating-importlib-import-module)has this block at the top of the function body:

```python
try:
    return sys.modules[absolute_name]
except KeyError:
    pass
```

Thus, it is impossible to do a proper deprecation cycle as long as the users uses the `import` statement.

**Comment by ezyang** [comment]
tagging @bdhirsh on this one too. Will need https://www.internalfb.com/diff/D26259045 (FB only) to go in first

**Review by bdhirsh — APPROVED** [review]

**Comment by bdhirsh** [comment]
Looks like there are still a few more tests failing internally. I'll try to make another patch for them first.

**Comment by bdhirsh** [comment]
Just fyi, I have a PR here https://github.com/pytorch/pytorch/pull/53139 that should help fix the other internal test failures, and probably help make this a tiny bit less BC breaking. Apparently some tensors have been serialized with `torch.save()` in such a way that `torch.load()`'ing them requires importing `torch.tensor`. So the metadata in the pickle file is telling pickle to `import torch.tensor` as part of the deserialization process. That PR adds hooks to automatically convert any `torch.tensor` imports to `torch._tensor` imports if it sees them.

**Comment by pmeier** [comment]
Superseded by #53424.
